### ISSUE TITLE
CI does not auto-rebase latest master changes.

### ISSUE THREAD

**Issue by pvalena** [sclorg/ci-scripts#51]
Reviewing PR sclorg/s2i-ruby-container#316, we've hit a corner-case which uncovered this issue.

### How to reproduce:

  - PR is created, on top of `master` branch.
  - In the meantime the master branch gets updated with non-conflicting changes (e.g. updating common scripts).
  - While reviewing the PR, issues are found, fix gets pushed and CI re-runs.
  - The CI gets the state of the current PR (with the `master` commits from that time), although there are additional non-conflicting changes.
  - Lack of the changes (some other fix) can cause the CI to fail, although it would pass when rebased.
  - The actual PR will get auto-rebased on merge.

### In current state:

  - The manual rebase serves no purpose, apart from actually running the CI with correct state.
  - In most cases, the PR will not get rebased with latest changes, and the CI result is not the latest one (e.g. missing some tests added in previous PR).
  - Therefore the results from the CI do not reflect the "final state", IOW state that will occur when the PR gets merged, as it was auto-rebased.
  - The very next PR (with no changes) can actually fail, as some breaking changes could be merged when an outdated test suite (or common scripts) was used for testing the previous PR.

 (Please to correct me if I'm mistaken.)

## Solution
I'd simply "auto-rebase" for the PR testing (rebase on latest origin/master), not just for the actual merge. This is the cleanest and most accurate solution. I did actually expect that behaviour, and I think it's a correct one.

In case when auto-rebase is not possible, the CI should not run either way.
I don't think the current state has any benefits. Are there some corner cases that I've missed?

Previously resiliency was listed as a reason to keep current state. This change would not do any difference in that state, as only the commits already in this repo are used. IOW `container-common-scripts` does not get auto-updated for CI testing, and any faulty change commited there will not be propagated, unless actually merged in this repo.

Bottom line, updating the 'common' commit (or any other change) is not a part of the PR, and always rebasing to the latest commit is not a requirement for merge as of now (because of the auto-rebase), unless the change is a conflicting one. F.e. gitlab has a `rebase` button on the PR, as it doesn't do auto-rebase the PR.

**Comment by pkubatrh** [comment]
>I don't have rights to move it to ci-scripts - can you do that ("Transfer issue" in the right panel).

Thanks for the pointer! I did not actually know you could move whole issues to a different repository

>I don't think that's the case. What you're describing are conflicting changes (more on that later), which can't be auto-rebased either way.

This is a bit offtopic from the rest of the issue, so I would not like to go too deep into this. But I feel more pain could be created by not following up on this, so here goes.
That is actually what you (and me) would think - it must be a conflicting change. Unfortunately, that is not the case.
Here is a simple example. I moved away from my original example with tests, since we do not have them copied into the versioned directories anywhere that I remember (we either use symlinks, or distgen), so the issue is not there anymore. But for changes in versioned files (eg. Dockerfiles), this can still be hit.
See the [PR](https://github.com/pkubatrh/s2i-ruby-container/pull/1) I created against my fork of ruby. The PR was originally created against this poc branch [commit](https://github.com/pkubatrh/s2i-ruby-container/commit/b847e52455c66df19d6ee10e6d9aa5be0d6ce957). After that a new poc branch [commit](https://github.com/pkubatrh/s2i-ruby-container/commit/31cee96d01a999bb972616dc5243b3c98f7896fc) was pushed, adding a new variable to `2.7/Dockerfile`. The PR was rebased (no conflicts). When you check the 2.7/Dockerfile file in the PR, you will see the new variable definition is no longer there.

>I don't think so. It's like saying auto-rebase is tricky, although github does it all the time (which proves it otherwise). There's no "magic" it should be fairly straightforward:

Oh the implementation of the actual auto-rebase feature would definitely be easy to achieve. I meant to say that the possible fallout from implementing the change might be harder to tackle. And that part is what needs the discussion.

>Even if it was really complicated situation, AFAICT what you're proposing is to keep in a worse state, because we don't have a "silver bullet". I still can't see any benefits of current state in comparison to my proposed enhanced state - please elaborate.

I am not proposing to keep it in the current state - as I said, I do not have a strong opinion on the matter. I do agree with you that there are no benefits I can see compared to your proposal. The question would then be more along the lines of "Does the proposal have enough benefits to spend the time on implementing it?"

>Sure, we can discuss, let's investigate if there are some corner-cases

Shared this with some other sclorg maintainers and am also planning to have a discussion on the next container image meeting - feel free to stop by!

>Well, if you really need the result, it should be marked as such, but the CI can fallback to run without the auto-rebase 

Yeah that would be nice to have. I do wonder how to implement it so that it is obvious that the test run succeeded/failed, but the PR should still not be merged.

**Comment by FaramosCZ** [comment]
Some ideas I want to share,
please correct me if I'm wrong with facts or assumptions.

---

1/ Block non-fast-forward pushes

When the push is no longer fast-forward, it meas, that in the meantime something changed on the remote.
In that moment you should rebase the PR. Rebase conflicts must be resolved in any case.
Should no rebase conflicts arise, the push will become fast-forward. No problem.

E.g. Fedora Pagure allows to block non-fast-forward pushes
( Settings -> Hooks -> Block non fast-forward pushes )

When you want make a push (accept PR) which is not fast-forward, it has to be done via "merge commit".
This can be IMO blocked on GitHub
( Settings -> Branches -> Require linear history "Prevent merge commits from being pushed to matching branches." ) 

Result?
For PR to be accepted, it has to be on top of the current target branch. That means CI will run on top of the latest content.

---

2/ Auto-rebase
Seems like a good idea to me.
In case no rebase conflict arises, everything's OK. In case a rebase conflict does arise, it must be resolved.
What might go wrong ? (like, really)

---

3/ Lot of PRs = tons of CI re-runs
If there is a lot of PRs actively being worked on, it is IMO right thing to do to require rebases and CI re-runs for all of them, everytime.
If there are lot of PRs, but only few really active; would it be possible to label PR as "stalled" or something like that to prevent auto-rebases & CI re-runs ?

---

If you really want to have CI and trust it, then there IMO aren't any cases you won't like it to run.
if you really don't want to run any portion of the CI there is something wrong with the design.
* Maybe you want to split your CI to tiers and run only certain tiers for certain cases ?
* Or maybe you want the CI split to suites, where each suite takes care of certain functionality area and usually you only need to pass that functionality area tests.

---

> In case when auto-rebase is not possible, the CI should not run either way.

Right

**Comment by phracek** [comment]
> Some ideas I want to share,
> please correct me if I'm wrong with facts or assumptions.
> 
> 1/ Block non-fast-forward pushes
> 
> When the push is no longer fast-forward, it meas, that in the meantime something changed on the remote.
> In that moment you should rebase the PR. Rebase conflicts must be resolved in any case.
> Should no rebase conflicts arise, the push will become fast-forward. No problem.
> 
> E.g. Fedora Pagure allows to block non-fast-forward pushes
> ( Settings -> Hooks -> Block non fast-forward pushes )
> 

I totally agree with this. 
> When you want make a push (accept PR) which is not fast-forward, it has to be done via "merge commit".
> This can be IMO blocked on GitHub
> ( Settings -> Branches -> Require linear history "Prevent merge commits from being pushed to matching branches." )
> 
> Result?
> For PR to be accepted, it has to be on top of the current target branch. That means CI will run on top of the latest content.
> 
> 2/ Auto-rebase
> Seems like a good idea to me.
> In case no rebase conflict arises, everything's OK. In case a rebase conflict does arise, it must be resolved.
> What might go wrong ? (like, really)
> 
It can be used. But tests have to pass.
We have daily tests per OS and including OpenShift.

> 3/ Lot of PRs = tons of CI re-runs
> If there is a lot of PRs actively being worked on, it is IMO right thing to do to require rebases and CI re-runs for all of them, everytime.
> If there are lot of PRs, but only few really active; would it be possible to label PR as "stalled" or something like that to prevent auto-rebases & CI re-runs ?
> 
> If you really want to have CI and trust it, then there IMO aren't any cases you won't like it to run.
> if you really don't want to run any portion of the CI there is something wrong with the design.
> 
>     * Maybe you want to split your CI to tiers and run only certain tiers for certain cases ?
> 
>     * Or maybe you want the CI split to suites, where each suite takes care of certain functionality area and usually you only need to pass that functionality area tests.
> 
> 
> > In case when auto-rebase is not possible, the CI should not run either way.
> 
> Right

**Comment by phracek** [comment]
We don't use ci-scripts in our CI anymore. Closing it.
